0.4.5: insertHour, path and polish

ArneBab opened the pull request:
This tracks my work and allows for easy review until the changes is merged. See the NEWS file for changes.

Commit:
syntax help, block elem first plus tabify

Commit:
parse description as textile and name content content. Thanks to banbananabans@CQUZ... from FMS

Commit:
also link to style.css at the bottom

Commit:
Add explicit path option

Commit:
also use the path for the link in the inserter and the site list

Commit:
footer: link to the latest version of Sharesite

Commit:
Update NEWS file

Commit:
ensure that getPath always has something to return

Commit:
url encode / in paths, because it breaks preview and insert.

Commit:
template: enclose description by div, not by p

Commit:
allow restricting inserts to a certain hour of the day

Commit:
improve UI strings

Commit:
fix source selection in Makefile

Commit:
allow setting the upload time to a random hour

Commit:
FIX: adjusted just the parameter, not the attribute

Commit:
Added signature for changeset 76c0e45f2bee

Commit:
Added tag 0.2.8 for changeset efb842bd4f73

Commit:
Added signature for changeset 7f938354de9c

Commit:
merge git tag and hg sign

Commit:
update NEWS

Commit:
add TODO for a random wait time.

ArneBab commented:
Do you know off-hand how to cleanly make the plugin wait up to 59 minutes without stalling anything? (see https://github.com/freenet/plugin-sharesite/pull/1/commits/80c16b603caf9dd63e31e7d92e65f8a715178672 )

xor-freenet commented:
> Do you know off-hand how to cleanly make the plugin wait up to 59 minutes without stalling anything? (see 80c16b6 )

What do you mean by "stalling" ? Are you running your inserts on the UI thread?
Run them on a thread of their own then.

For such purposes, the BackgroundJob and especially DelayedBackgroundJob had recently been added to WoT:
https://github.com/freenet/plugin-WebOfTrust/tree/master/src/plugins/WebOfTrust/util/jobs

xor-freenet commented:
(The BackgroundJob stuff is very generic, I'd be willing to file a PR to add it to fred if stuff other than WoT starts using it as a copy-paste)

ArneBab commented:
xor-freenet writes:

> > Do you know off-hand how to cleanly make the plugin wait up to 59 minutes without stalling anything? (see 80c16b6 )
> 
> What do you mean by "stalling" ? Are you running your inserts on the UI thread?
> Run them on a thread of their own then.

Frankly put, I don’t know its internals - and I want to keep the code
operating in a way which allows hacking it without learning the
internals.

With stalling I mean: I do not know whether putting a sleep in that
thread could block any other part of fred. If not, I’d simply put a
sleep there.

A running insert does not block the UI of Sharesite.

Best wishes,

## Arne

Unpolitisch sein
heißt politisch sein
ohne es zu merken

Commit:
Makefile: Always run ant clean before ant dist

Commit:
get the version of the running plugin instance, not the plugin class.

Commit:
add a random wait time within the selected hour interval to mask triggering an insert in that hour

Commit:
public static final -> private static final *happy*

Commit:
Increase version to 0.4.0: We now support convenient anonymous usage

Commit:
update NEWS

Commit:
textilej: add a special case for freenet keys (thank you, you know who you are)

Commit:
file bugs to keep the feature requests from FMS

Commit:
add a template for the initial content (preparation to ease starting to write)

Commit:
show insertHour setting after the path (at the top)

Commit:
version 0.4.1

Commit:
insertHour below content field

Commit:
add default content

Commit:
update NEWS

Commit:
Added tag 0.4.1 for changeset 2a59727908b1

Commit:
Added signature for changeset 7d4d68e90ef7

Commit:
switch to Java 1.7 as source and target

Commit:
use try-with-resources to avoid resource leaks. Thanks to Bombe\!

Commit:
cleanup whitespace

xor-freenet commented:
Is it possible that you're accidentally pushing lots of unrelated code to this branch?
The recent commits don't seem to be related to the description of the PR:

> This adds insertHour (random by default) and path, so you can hide your activity profile and conveniently set a nice title.

ArneBab commented:
Uh, yes, I do. Since this was not merged, I changed the title of the pull-request accordingly.

Commit:
add show current utc time bug

Commit:
Show the current UTC hour next to the InsertHour field

Commit:
FIX: used wrong replacement value for current utc hour

Commit:
update to 0.4.2

Commit:
remove unused variable, thanks to Bombe

Commit:
Added tag 0.4.2 for changeset df61855bf1ad

Commit:
Added signature for changeset 7e2afe9b55ca

ArneBab committed:
increase realVersion and version

Commit:
ant: add a copy task to provide a Sharesite.jar for the release scripts

ArneBab committed:
0.4.4 with updated ant

Commit:
fix regression: busy wait for queued websites

Commit:
Added tag 0.4.5 for changeset 792153e157d9

Commit:
Added signature for changeset 63ac7f7777df

Bombe commented on src/net/java/textilej/parser/markup/textile/phrase/SimpleTextilePhraseModifier.java:
```diff
@@ -49,7 +49,15 @@ public SimpleTextilePhraseModifier(String delimiter, SpanType spanType, boolean
-		
+
+		// special casing to avoid breaking freenet keys:
+		if (this.delimiter == "-") {
```
Nope.

ArneBab replied:
how would you do it?

Bombe replied:
Strings have to compared using `.equals()` (except for some very constrained circumstances which do not apply here), so `delimiter.equals("-")` would be correct. If `delimiter` can only be `null` you’d have to check for that, too, because (obviously) you can’t call `.equals()` on `null`.

Bombe replied:
Strings have to be compared using `.equals()`, so `if (delimiter.equals("-"))` would be correct—unless `delimiter` can be `null`.

Bombe commented on src/plugins/Sharesite/Freesite.java:
```diff
@@ -70,7 +76,22 @@ public Freesite(int uniqueKey) {
+			 ) {
+			StringBuilder sb = new StringBuilder();
+			while (true) {
+				String line = reader.readLine();
```
Why are you parsing this into lines when…

Bombe commented on src/plugins/Sharesite/Freesite.java:
```diff
@@ -70,7 +76,22 @@ public Freesite(int uniqueKey) {
+			}
+			
+			reader.close();
+			this.text= sb.toString();
```
…all you do with it is to put into a single `String` again?

ArneBab replied:
That’s just because I did not see a way to simply drain the whole reader into a string (and java kept telling me I’m doing it wrong — which likely was true). I’ll *very* happily use an approach with less line-noise here.

Bombe commented on src/plugins/Sharesite/Freesite.java:
```diff
@@ -90,6 +111,35 @@ public synchronized void setName(String name) {
+	}
+
+	public synchronized void setInsertHour(Integer insertHour) {
+		if (insertHour.equals(-2)) {
```
Don’t use `.equals()` here, it will fail when you hand in `null`. If you don’t ever have to hand in `null`, turn the parameter into an `int`.

ArneBab replied:
how can I make null parse to -2? (that would be the best action: in case of error choose the safe option)

Bombe replied:
I’m not sure what you mean here. You’re not parsing `insertHour` into anything and I have no idea whether `insertHour` being `null` is even a valid scenario here. If you want the code to treat `null` the same as `-2`, just check for both: `if ((insertHour == null) || (insertHour == -2))`.

ArneBab replied:
Oh, yes, it’s that easy. I keep forgetting that java is pretty low-level on this (not meant as derogative: it feels high level most of the time but can suddenly drop very low and allows writing efficient code on a pretty low level — something I would not dare to do in Python).

insertHour == null could happen if there’s something wrong with the storage.

Bombe commented on .hgsigs:
```diff
@@ -3,3 +3,8 @@ be98370f35044c24e60f75102ff26091d4614864 0 iQEcBAABAgAGBQJWyH6YAAoJEKMv031G/7TTm
 038bde8cbc27552c1293aa7020e6eb39236f2209 0 iQEcBAABAgAGBQJWyKd4AAoJEKMv031G/7TT24EH/1D3Pc7e0py3W7nkvu569xaqQfZje62ipMahZPi4JPSBuMWZpQnaoT1e9NASsiWgJQtsPj9/heNn05sT7Pwxcg0G1VeilZqpjKmvgVGt3txeus+IJVQ0X4cpe8FY04+VrQ1SkCyncZbeQTMU7b5504ic8hr4F3F7HN6IW7Kms+2hEgnEOkrJJkjPO8TJZ1OFgowSQSKA8QbJKI7Fp+dNyMpj42TM+0zcGk4e23yKYsQix085YM6tNgLA+yHg6nhBg5niW7Anzr6DMjyYRDrSiveZMPkC/0194wiDz3UWfW9qsoroJUFIpAmcwwpP ...
 cd12c2e21dad7034b6775e3ed3d2d8be1ab46a73 0 iQEcBAABAgAGBQJW4eAhAAoJEKMv031G/7TTW0cH/jqSHwWFsjnynliXCnXY5W+HBUYQ2+KIJN7swAgjS4Wjn5OBrbKxg6TNtlCiCsMAIGvgXFvD+G8R6wtVziL8is6cyzK++tLdFUUOlfQJuvxAhCMYJlf+qKwuhoXH0M/m9Ff0+AFgaBdPLSDLWxIHLM+nWdKFMBlbpx+Pbt9JLz2CNWJax4+ULsXqFQUd++rcL3WqRWHROFiP6X9XqrpNwET6bUHyvg5F9CMO+5JxKuWnIYN1/zJ5j3rZi4vgjNPUgkgWWbHHtuUzHFP2IJfj+jI4OcRZoLv/i4Mjk1xJAWB61P0AWfhkgGK9XpTr ...
 e3c31f3768f3a5a5cc0f88201d3b93d4fa3d9245 0 iQEcBAABAgAGBQJXMv+PAAoJEKMv031G/7TTji4H/3UNRufP7wvmmeUBFB59s+atvG6TSeH3twUe9+x2+jpFBXt8E96fDWYD1kcuOf2rSYQ0JeN+munOjjhQdldOBUIt8C5ww+Y1HYq53QaA9QiUF7IAMj1tRYQ+rtF1/5RfGCMHGWIeCTIx/ckHnnvnLLeaCVPIvQzoN1ATGmOe/1Ojz10Aze+2Wk6evAYJv4WqUWCai/zugIZIskfRNFrVKXOYNF2kTPogOFIPO0Q7K0anknl1HszeW+U/vJXRoWXVzgRMexg/G+i8MRfsySL265FWD/J1u3DtC3rTw/khDJBImdVMR2vnQu/W4p6M ...
+76c0e45f2beed9a7cfd4a04aa271554a10a7b8df 0 iQEcBAABAgAGBQJX0WZ6AAoJEKMv031G/7TTT/QIAL8bbkMB4cenAJMfX0r0Q+OinkfTu4TYjz3HvM91aookeF2bDHF0vwYnQvqH1IxL3mK4fMcks9RTcGTl2CLZix6iSMDWeYLMWOK1qHFncvP/Ff9rDZZvOxwqABRGPO2bv/6mIG0Agr6aIxt3iwetWq0v++mwvsBQ7DTO9oOXbKhVfYx+/KYdWekcSeBIrxI6znXoR3IOsYH/7rKKJK1Pf/v7QBUD9nrDt3hbUhvWpdZElJaLr6/dECZ2Afaaii5BfgNB1kmc0nQkR2AkvKIVMnt7dXQkKomxkCepFfTT+s6IyAkx4x8h87OV/Y0H ...
```
Where is there mercurial-specific stuff in this repository?

ArneBab replied:
because the git repo is the secondary repository while the primary one lives on Freenet.

Bombe commented on Makefile:
```diff
@@ -6,7 +8,8 @@ check-syntax: src
 
-dist/: src build.xml build.txt
+dist/: src build.xml build.txt $(SOURCES)
+	ant clean
```
This is so wrong. :(

ArneBab replied:
This part actually is wrong (it should only rebuild if there actually were changes — fixed now).

I assume that you also mean including a Makefile at all, but since Makefiles are the one cross-language standard for defining how to build a program, I disagree on this. I wish we had Makefiles in fred to drive gradle, so people who are not used to gradle would not have to hunt for the compile commands¹.

A plain Makefile would make switching build systems much less painful: just swap a line executing ant with a line executing gradle, and all which commandline users notice would be that the tests are faster and the build output cleaner. (But I won’t press this philosophical discussion).

Anyway: Thank you for your review!

¹: when I wanted to do quick experiments in next that lost me much more time than I’m comfortable losing to a tool which should only turn a small change into something I can test and run

Bombe commented on src/plugins/Sharesite/Inserter.java:
```diff
@@ -31,27 +34,47 @@ public Inserter() {
+				// TODO: Wait for a random fraction of the remaining
+				// part of the current hour to prevent detection of
+				// people who click insert during the insertHour.
+				if (!nextToInsert.getInsertHour().equals(-1)) { // no instant insert
```
`Integer`s don’t have to be compared with `.equals()`, the VM automatically boxes/unboxes as required.

Bombe commented on src/plugins/Sharesite/Inserter.java:
```diff
@@ -31,27 +34,47 @@ public Inserter() {
+						// need to back to ensure we do not miss a slot
+						if (nextToInsert == null) {
+							// empty run, can sleep until next full hour
+							Thread.sleep((60 - currentMinute) * 60 * 1000);
```
I really hope this thread is not called from any UI anywhere.

ArneBab replied:
It’s the main run loop of the plugin and should never be called by the user (else we’d have to infinite loops competing for the resources).

Bombe commented on src/plugins/Sharesite/Inserter.java:
```diff
@@ -60,6 +83,18 @@ public void run() {
 	}
 
+	private synchronized Freesite getNextToInsert(Integer currentHour, LinkedList<Freesite> queuedInserts) {
+		for(Freesite c : queuedInserts) {
```
Why is the variable that contains a `Freesite` object called `c`?

ArneBab replied:
it’s the *c*urrently checked Freesite in a tight loop (so the meaning of the variable should be clear). And it’s done like that in other places in the plugin. Is that overly surprising to java devs? If yes, what would you suggest?

Bombe commented on src/plugins/Sharesite/Inserter.java:
```diff
@@ -60,6 +83,18 @@ public void run() {
+	private synchronized Freesite getNextToInsert(Integer currentHour, LinkedList<Freesite> queuedInserts) {
+		for(Freesite c : queuedInserts) {
+			if (c.getInsertHour() == null
+				|| c.getInsertHour().equals(-1)
```
See above for using `.equals()` on `Integer`s.

Commit:
Makefile: avoid rebuilding without changes

Bombe commented on src/plugins/Sharesite/webui/EditToadlet.java:
```diff
@@ -56,13 +59,14 @@ public void handleMethodGET(URI uri, HTTPRequest req, ToadletContext ctx) throws
 		PageNode pageNode = pageMaker.getPageNode(l10n.getString("Sharesite.Menu.Name"), ctx);
 		HTMLNode editForm = pr.addFormChild(pageNode.content,"/Sharesite/Edit/" + siteId, "editForm");
-		addNodes(editForm,c.getName(),c.getDescription(),c.getText(),c.getCSS(), c.getActivelinkUri(),c.getRequestSSK(),c.getInsertSSK());
+		String iHour = Integer.toString(c.getInsertHour());
```
How does `Integer.toString(x)` behave when `x` is `null`?

ArneBab replied:
getInsertHour never returns null: if insertHour is null, getInsertHour returns -1. Which is not what I want, but is in there for backwards compatibility (to avoid complaints from users that their sharesites suddenly do not insert instantly anymore). Can’t avoid that since there already are several users …

Commit:
Makefile: check and check-syntax dependencies

They should depend on SOURCES, not just the state of src/

Bombe commented on src/plugins/Sharesite/webui/EditToadlet.java:
```diff
@@ -114,62 +126,76 @@ private void addNodes( HTMLNode form, String name, String desc, String text, Str
 		vals = new String[] { "textInput", "20", "80", "font-size: medium;" };
 		textDiv.addChild("textarea", attrs, vals, text);
 
+		HTMLNode insertHour = editBox.content.addChild("p");
```
I think it’s ill-advised to call an HTMLNode “insertHour” while the actualy insert hour is called “iHour” which looks like a variable with the Hungarian prefix for an integer but is actually a `String`. That’s, like, maximum surprise.

ArneBab replied:
that’s true … it’s been too long since I’ve seen type-prefixes. And there is no reason to disambiguate this here (which is what the surrounding code does with things like aUri), so I can change this without making the style inconsistent. Thank you!

Bombe commented on src/plugins/Sharesite/webui/EditToadlet.java:
```diff
@@ -250,21 +274,28 @@ public void handleMethodPOST(URI uri, HTTPRequest req, ToadletContext ctx) throw
 			String name = req.getPartAsStringFailsafe("nameInput", 1000).trim(); // 1kB
+			String path = req.getPartAsStringFailsafe("pathInput", 1000).trim(); // 1kB
 			String desc = req.getPartAsStringFailsafe("descInput", 10000).trim(); // 10kB
 			String text = req.getPartAsStringFailsafe("textInput", 1000000000).trim(); // 1GB
```
Seriously? 1 *GIGA*​byte?

ArneBab replied:
not my code :) (where is the "hand waving, I’m not touching this and hope there’s a reason"-emoji?)

Anyway: This is strange, but not problematic, since it would only be the user hurting him/herself (freenet should handle uploading 1GB of data).

Commit:
add missing checks for null values

Commit:
refactor: rename variable iHour to plain hour

Commit:
add bug

xor-freenet commented:
@ArneBab:
I would really like this monster-PR to go away in favor of small ones which obey our workflow of "1 PR = 1 thing", please stop pushing more code to it.

Remember you were one of the people who confronted me with the strongest criticism for the size of the plugin-fcp-rewrite PR ;)

ArneBab commented:
I would also like this PR to go away — by being merged. I’m not pushing to the main repo yet because I did not get any feedback on the fixes for the review comments. If this is ok-to-release, this pull request will be merged and closed quickly.

xor-freenet commented:
@ArneBab 
> I would also like this PR to go away — by being merged.

Well what I'm trying to say is: If you would instead submit the work as reasonably small PRs then they would have a much higher probability of being merged than this giant one.

> I’m not pushing to the main repo yet because I did not get any feedback on the fixes for the review comments. If this is ok-to-release, this pull request will be merged and closed quickly.

Didn't you already release this stuff as a binary to the main network anyway?
Our repositories should match the binaries on the network!

ArneBab commented:
xor-freenet <notifications@github.com> writes:
> Well what I'm trying to say is: If you would instead submit the work as reasonably small PRs then they would have a much higher probability of being merged than this giant one.

This one is already mostly reviewed (only the fixes for the
review-comments haven’t been) and cutting it apart would be
much more work than the actual improvements I want to do — and would
destroy the review discussion.

IIRC I did not yet deploy the improvements here as actual plugin update,
only as testing jar outside our plugin mechanism.

Commit:
add syntax help for the audio tag

Commit:
content-template: add a tip against style analysis

Commit:
break run-loop when interrupted

Commit:
Inserter: actually *break* when interrupted

Commit:
Inserter: whitespace cleanup

Commit:
Prepare for 0.4.6

Commit:
Use a ticker instead of a while loop.

Probably doing it wrong, but it seems to work.

Commit:
merge useticker branch

Commit:
tick within 10ms after requesting an insert

Commit:
update version to 0.4.7

Commit:
fix: undid my breaking change to the correct pull-request of TheSeeker

ArneBab committed:
build.xml: adapt to new jarpath of gradle-based fred

ArneBab committed:
more beautiful default style + example for header image using the activelink.

Commit:
new default css thanks to w7Q2J@QTPnftHRW6dAF… from FMS

Readability increases a lot, and it seems more friendly while keeping
it simple.

Commit:
merge
